slides/E-async-web.md

```diff
@@ -0,0 +1,892 @@
+
+---
+layout: default
+---
+
+# `async`/`await` expansion takeaways
+
+- Rust generates state machines out of `await`s
+- Generated code may become very complex, but original is easy to follow
+
+---
+layout: section
+---
+
+# Running `Future`s
+
+---
+layout: default
+---
+
+# What's an `async` Runtime do?
+
+- Spawn `Future`s
+- Keep track of pending `Future`s
+- Call `Future::poll` on each `Future` that can make progress
+- Poll `Future`s on `Waker::wake` calls
+
+Nice to have:
+- Poll `Future`s on multiple threads
+- Abstract over I/O
+
+*Crates depending on different runtime I/O abstractions be incompatible!*
+---
+layout: default
+---
+
+# Many Runtime flavors
+
+- [`smol`](https://github.com/smol-rs/smol): Small
+- [`async-std`](https://async.rs/): API resembles `std`
+- [`tokio`](https://tokio.rs): Bery commonly used
+- [`embassy`](https://embassy.dev/): Embedded
+- Create your own?
+
+*Note: crates may depend on a specific runtime!*
+
+---
+layout: default
+---
+
+# Showcase: Tokio
+
+```rust
+/// Set up a tokio Runtime and spawn the Future returned by `main`
+#[tokio::main]
+async fn main() {
+    do_stuff.await();
+}
+```
+
+It does stuff!
+
+---
+layout: default
+---
+
+# A simple TCP server
+
+```rust
+use tokio::net::{TcpListener, TcpStream};
+
+/// Read a line, and reply with that line!
+async fn handle_connection(socket: TcpStream) -> anyhow::Result<()> {
+    let mut stream = BufReader::new(socket);
+    let mut name = String::new();
+    stream.read_line(&mut name).await?;
+    
+    stream.write_all(format!("Hello, {name}!").as_bytes()).await?;
+    Ok(())
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let listener = TcpListener::bind("127.0.0.1:6379").await.unwrap();
+
+    loop {
+        // The second item contains the IP and port of the new connection.
+        let (socket, _) = listener.accept().await.unwrap();
+        handle_connection(socket).await?;
+    }
+}
+```
+
+---
+layout: default
+---
+# It works!
+
+```bash
+$ echo -e Ferris | nc localhost 6379
+Hello Ferris!⏎
+```
+
+*Question: But does it scale?*
+
+<v-click>
+<div>
+Nope! Only one request at a time!
+</div>
+</v-click>
+
+---
+layout: default
+---
+
+# Spawning tasks is cheap!
+```rust
+#[tokio::main]
+async fn main() -> Result<()> {
+    let listener = TcpListener::bind("127.0.0.1:6379").await.unwrap();
+
+    loop {
+        // The second item contains the IP and port of the new connection.
+        let (socket, _) = listener.accept().await.unwrap();
+        handle_connection(socket).await?;
+    }
+}
+```
+becomes:
+```rust
+async fn main() -> Result<()> {
+    loop {
+        // The second item contains the IP and port of the new connection.
+        let (socket, _) = listener.accept().await.unwrap();
+        tokio::task::spawn(async {
+            handle_connection(socket).await?;
+            Ok::<_, anyhow::Error>(()) // <-- cool trick to specify error type
```

Comment:
In this slide my eyes are drawn to this comment, but I think the focus should be on the `spawn` part. (And I don't think it's cool that we *need* the cool trick, but that can't be helped I guess)